packages/pretty-format/src/plugins/lib/markup.ts

```diff
@@ -62,14 +65,23 @@ export const printChildren = (
   printer: Printer,
 ): string =>
   children
-    .map(
-      child =>
-        config.spacingOuter +
-        indentation +
-        (typeof child === 'string'
+    .map(child => {
+      const printedChild =
+        typeof child === 'string'
           ? printText(child, config)
-          : printer(child, config, indentation, depth, refs)),
-    )
+          : printer(child, config, indentation, depth, refs);
+
+      if (
+        printedChild === '' &&
+        typeof child === 'object' &&
+        child !== null &&
+        (child as Node).nodeType !== NodeTypeTextNode
+      ) {
+        // A plugin serialized this Node to '' meaning we should ignore it.
```

Comment:
Hmm, I'm not so excited about taking away the ability of plugins to output empty text for a node, esp. in a breaking way. Would it be an option to allow returning `null` instead?

Reply:
> Hmm, I'm not so excited about taking away the ability of plugins to output empty text for a node, esp. in a breaking way. 

My question would be: Why is that desired? Why write a plugin that takes a Node, serializes it to an empty string but also expect that empty string to be included in the snapshot.

> Would it be an option to allow returning null instead?

I've tried this route but it propagated the possibly `null` return all the way up so that other packages started to break.
I'll check again if we can change the boundaries but so far it seemed there are too many places that are accidentally coupled.

Reply:
I'm not saying I have a concrete use case, but I know how many usages are out there pretty-printing element trees that are abstractions over HTML (if you need a horrifying example, take a look at 'ExtReact'). If in just one of them one renders plain text not with string children, but with `createElem('Label', {text: ''})`, and their pretty printer converts the custom component language to the actual HTML, it will now no longer print the right representation of the element tree. With something used in as many wildly different tech stacks as Jest is, it is bound to break *someone*.
I would follow principle of least surprise — if an actual `''` text child is printed as an empty (except indented) line, you would expect a child that is mapped to `''` by the printer to behave the same. I don't know the empty string to typically signalize omission in other contexts either.

Regarding an explicit omission return value, we would want that limited specifically to `printChildren`, because while printing children you can omit one, not in all other contexts.
I would expect that changing the type of `printChildren`s `printer` argument to something like the following would work?
```ts
type NullablePrinter = (...args: Parameters<Printer>) => ReturnType<Printer> | null
```

Reply:
Actually, I've just looked at the way you implemented your plugin, and this would not really help you because you are not in control of the `printChildren` call, instead you rely on the standard `DOMElement` or other plugin to make that call, passing the standard printer along which will then delegate to your plugin. :/

Reply:
I believe we have both been thinking about this the wrong way.

Printing
```
<div>
  ${print('asdf')}
  ${print('')}
</div>
```
and expecting the `print('')` to be able to remove the whole third line is like printing
```
Object {
  a: ${print(42)}
  b: ${print(null)}
}
```
and expecting the `print(null)` to be able to remove the whole third line for a `toEqualIgnoringNulls()` matcher.

`print('')` at top level should have the same result as `print('')` in an element child, and `print('')` at top level does not return a zero-line string either (because a zero-line string is not a thing, it would have -1 line breaks in it).

I think what should really be done to omit children from the output is not pass them into a printer in the first place. Replace the DOMElement plugin with a plugin under your control, that clones the node, throws out any children you don't want, then pass it to the DOMElement plugin to reuse its code. Let me know if this works for you.

Reply:
> Actually, I've just looked at the way you implemented your plugin, and this would not really help you because you are not in control of the

Did you see the test added?

Reply:
Yes, the sentence you quoted referred to the previous comment, not to the implementation in this PR.

Reply:
I don't understand why you can say the current implementation doesn't help me when the test that uses the new implementation is exactly my desired outcome.

Reply:
Again, I am not saying what you implemented here does not work for you, I am aware it does. The second of the three comments says that what I suggested in the second half of the first comment actually would not work for you after I reconsidered it.
For why this change will lead to surprising behavior, see the first half of my first comment.
For an alternative solution that I believe would work without introducing this surprising behavior, see my third comment.